frontend/src/containers/App.js

```diff
@@ -0,0 +1,194 @@
+import React, { useEffect, useRef } from 'react';
+
+import * as rmsClient from '../api/RoomzMediaServerClient.js'
+import './App.css';
+
+
+
+function App(props) {
+  const room_id = useRef();
+  const user_id = useRef();
+  const egress_media = useRef();
+
+  var ICE_SERVERS = [
+    {urls:"stun:stun.l.google.com:19302"}
+  ];
+  var roomyPcs = {};
+  let videos = [
+    { peer_id: "0-0" },
+  ];
+  let videoRefs = [];
+  var myPeerId = "";
+  let egressMediaStream = null;
+
+  useEffect(() => {
+    rmsClient.askToConnect()
+  })
+  
+  // setupLocalMedia requests access to the user's microphone and webcam and
+  // properly sets up the egress media stream.
+  // NOTE: This will likely be called on load within the vestibule component.
+  function setupLocalMediaUtil(cb, eb) {
+    if (egressMediaStream != null) {
+      if (cb) cb();
+      return
+    }
+    console.log('asking for local audio/video inputs')
+    navigator.getUserMedia = (navigator.getUserMedia ||
+      navigator.webkitGetUserMedia ||
+      navigator.mozGetUserMedia ||
+      navigator.msGetUserMedia);
+    
+    // TODO: Pass config to mute audio/video.
+    navigator.getUserMedia({"audio": true, "video": true},
+      function(stream) {
+        console.log('granted access to audio/video')
+        egressMediaStream = stream
+        egress_media.current.srcObject = egressMediaStream
+        if (cb) cb();
+      },
+      function() {
+        console.log('access denied for audio/video')
+        alert('have fun being lame on zoom')
+        if (eb) eb();
+      });
+  }
+
+  function setupLocalMedia() {
+    setupLocalMediaUtil(() => {
+      console.log('successfully setup local media')
+    })
+  }
+
+  // joinMediaRoom emits the 'JoinMediaRoom' event to the RMS and registers
+  // event handlers for possible response events from the RMS.
+  function joinMediaRoom() {
```

Comment:
This function is quite large, ideally it would be split up. Totally understandable that you are still experimenting and trying to get it functionally working first.

Reply:
I'll split up in the future